### PR TITLE
Inline during parsing

### PR DESCRIPTION
**Pull request by marjanamatf**

### CODE REVIEW

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageGenerator.java [hunk 1]
```diff
@@ -1113,6 +1107,13 @@ public static void registerGraphBuilderPlugins(FeatureHandler featureHandler, Ru
         SubstrateReplacements replacements = (SubstrateReplacements) providers.getReplacements();
         plugins.appendInlineInvokePlugin(replacements);
 
+        // if analysis is false, don't add plugin
+        if (analysis && NativeImageInlineDuringParsingPlugin.Options.InlineBeforeAnalysis.getValue()) {
```
Comment (marjanamatf) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageGenerator.java:1120]: @vjovanov
 ne dodaje se ako je analysis false (bez ovoga je ukljucivan i nakon analize pa se dolazilo do Hosted metoda greskom)

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+        if (providers.getGraphBuilderPlugins().getInvocationPlugins().lookupInvocation(method) != null)
+            return null;
+
+        if (((AnalysisMethod)method).buildGraph(b.getDebug(), method, providers, Purpose.ANALYSIS) != null)
+            /* Method has a manually constructed graph via GraphProvider. */
+            return null;
+
+        // method is declared in user's class
+        if(SubstrateOptions.Class.getValue().equals(method.format("%H"))) {
```
Comment (marjanamatf) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: @vjovanov 
izdvojene su za sada samo metode koje su korisnicki definisane, da se za njih dobije graf

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                System.out.print(node.toString());
+                if (node instanceof LoadFieldNode)
+                    System.out.print(" - node represents a read of a static or instance field.");
+                if (node instanceof StoreFieldNode)
+                    System.out.print(" - node represents a write to a static or instance field.");
+                if (node instanceof ParameterNode)
+                    System.out.print(" - node represents a placeholder for an incoming argument to a function call.");
+                if (node instanceof ConstantNode)
+                    System.out.print(" - node represents a constant");
+                System.out.println();
```
Comment (marjanamatf) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: @vjovanov
uspesno se parsira graf za metod, za sada su samo ispisani cvorovi

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import static com.oracle.graal.pointsto.infrastructure.GraphProvider.*;
+
+public class NativeImageInlineDuringParsingPlugin implements InlineInvokePlugin {
+
+    // to use this plugin: native-image Example -H:+InlineBeforeAnalysis
```
Comment (marjanamatf) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java:73]: @vjovanov
promenjen fleg za ukljucivanje plugin-a

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageGenerator.java (hunk 1 above)
Comment (vjovanov) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageGenerator.java:1120]: Yup, good observation.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 2 above)
Comment (vjovanov) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: I would use the code from Christian here. He already went through the whole filtering process.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java (hunk 3 above)
Comment (vjovanov) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java]: That is awesome. Now, we can start detecting the simplest methods.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+    public NativeImageInlineDuringParsingPlugin(boolean analysis, HostedProviders providers) {
+        this.analysis = analysis;
+        this.providers = providers;
+    }
+
+    @Override
+    public InlineInfo shouldInlineInvoke(GraphBuilderContext b, ResolvedJavaMethod method, ValueNode[] args) {
+       /* InvocationData data = ((SharedBytecodeParser) b).inlineInvocationData;
+        if (data == null) {
+            throw VMError.shouldNotReachHere("must not use SubstrateInlineDuringParsingPlugin when bytecode parser does not have InvocationData");
```
Comment (marjanamatf) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java:211]: it's a problem: throws VMError while parsing some methods, for example com.oracle.svm.core.graal.jdk.ArraycopySnippets.doArraycopy

Comment (vjovanov) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java:211]: `doArraycopy` is a snippet and it should never end up being inlined in the first place. We have to filter it earlier.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java [hunk 6]
```diff
@@ -190,10 +193,22 @@ public NativeImageInlineDuringParsingPlugin(boolean analysis, HostedProviders pr
...
+            // System.out.println(((AnalysisMethod) (b.getMethod())).getDeclaringClass().getJavaClass());
+            return null;
+        }
+
+        if (Snippets.class.isAssignableFrom(((AnalysisMethod)method).getDeclaringClass().getJavaClass())) {
+            /* We are not interfering with any snippet */
+            // System.out.println(((AnalysisMethod)method).getDeclaringClass().getJavaClass());
+            return null;
+        }
+
```
Comment (marjanamatf) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java:208]: > `doArraycopy` is a snippet and it should never end up being inlined in the first place. We have to filter it earlier.
@vjovanov is it ok to filter snippets this way?

Comment (marjanamatf) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java:208]: after adding this filter, null InvocationData continues to be a problem for some methods from packages com.oracle.svm.core.snippets and com.oracle.svm.core.genscavenge  - can I filter them the same way?

**Pull request closed without merging**